src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineCloud.java

```diff
@@ -162,21 +173,20 @@ public void addConfiguration(InstanceConfiguration config) {
                     public Node call() throws Exception {
                         long startTime = System.currentTimeMillis();
                         LOGGER.log(Level.INFO, String.format("Waiting %dms for node %s to connect", config.getLaunchTimeoutMillis(), node.getNodeName()));
-                        while ((System.currentTimeMillis() - startTime) < config.getLaunchTimeoutMillis()) {
-                            LOGGER.log(Level.INFO, String.format("%dms elapsed waiting for node %s to connect", System.currentTimeMillis() - startTime, node.getNodeName()));
-                            Computer c;
-                            try {
-                                c = node.toComputer();
-                                if (c != null) {
-                                    c.connect(false).get();
-                                }
-                            } catch (Exception e) {
-                                LOGGER.log(Level.WARNING, String.format("Exception waiting for node %s to connect", node.getNodeName()), e);
+                        try {
+                            Computer c = node.toComputer();
+                            if (c != null) {
+                                c.connect(false).get(config.getLaunchTimeoutMillis(), TimeUnit.MILLISECONDS);
+                                LOGGER.log(Level.INFO, String.format("%dms elapsed waiting for node %s to connect", System.currentTimeMillis() - startTime, node.getNodeName()));
+                                return null;
+                            } else {
+                                LOGGER.log(Level.INFO, String.format("No computer for node %s found", node.getNodeName()));
                             }
-                            return node;
+                        } catch (Exception e) {
+                            LOGGER.log(Level.WARNING, String.format("Exception waiting for node %s to connect", node.getNodeName()), e);
                         }
-                        LOGGER.log(Level.WARNING, "Failed to connect to node within launch timeout");
-                        return node;
+                        node.terminate();
```

Comment:
This should be checking if the instance is a one shot correct? Otherwise we still want to keep it alive.
Separately I'm curious about the timing of the node termination here, do we want to terminate right after provisioning or after the job being run on the node has completed?

Reply:
That termination is just in case that node was not "provisioned". (there was exception)
In normal case we are terminating it in `OnceRetentionStrategy`..

Reply:
If this is only expected to happen if there was an exception, then the call to terminate should be within the catch block to make that clear.